Feat/algo-dev

cgrinaldi opened the pull request:
I've implemented the first version of the learning algorithm. It ended up being easier to use the algorithm that determines weights for a user's problems, and then samples from them. The learningAlgo service now exposes the method getProblem() instead of getProblems().

At a high-level, the algorithm calculates a weight using: Weight = (Age Component) \* (Progress Component) \* (Effort Component). Here is additional info about each of them:
- Age Component: Looks at how long it has been since the user has seen a problem
- Progress Component: Looks at the average rating the user has given a problem for their most recent X number of attempts
- Effort Component: Looks to see if the user's last attempt on the problem was a failure.

Once the weights are calculated for all of the users' problems, the following steps occur:
- The weights are normalized (which means their sum equal 1)
- The problems are sorted from least weight to greatest
- A [CDF](http://en.wikipedia.org/wiki/Cumulative_distribution_function) is calculated
- A random number is selected using `Math.random()` in order to determine which problem has been sampled from the distribution. 

**NOTE**: This pull request does step slightly into both of your domains, so there might be a merge conflict, but it should be slight:
- In userModel.js, I changed the dummy data and make the assumption that the user rating will be a number between 0 and 1 (**TODO**: Garrett, can you make sure to update the schema accordingly when you get to it?)
- In editorController.js, I updated the code to use the getProblem() instead of getProblems()

Finally, it should be noted that when a call to getProblem() is made, that problem is removed from the problems array so that the user won't see the same problem in the same session. If no problems are remaining, a value of null is returned from getProblem().

cgrinaldi committed:
(feat) Add helper function for millisecond to days conversion

cgrinaldi committed:
(fix) Change convertMStoDAYS to function declaration and round

cgrinaldi committed:
(feat) Create calcAgeWeight

Function takes in a given problem and calculates the associated
weight based on how many days it has been since the most recent
attempt.

Currently, the key assumption is that the most recent attempt
will be the last item on the attempts array.

cgrinaldi committed:
(feat) Create calcAverageRatings helper function

Function takes a problem's attempts and the number of most recent
attempts to be averaged across.

This value will be useful for the calcProgressWeight function.

cgrinaldi committed:
(feat) Return progressWeight value and change divisor for avg. rating

Before, we were dividing by the number of ratings present. However,
this will be biased when # ratings is less than numToAvg since if they
do well on the first problem, they are unlikely to see it.

cgrinaldi committed:
(doc) Add a TODO RE: Data Model

cgrinaldi committed:
(feat) Create calcEffortWeight

Function returns 1 of two values. 1 if the user gave a non-zero
rating for their most recent attempt (so they didn't completely
miss it), or 10 if they missed their last attempt.

cgrinaldi committed:
(feat) Create calculateWeights()

Function takes an array of problems (stored in the userData) and
will return an array of tuples where tuple[0] = weight and
tuple[1] = problem.

cgrinaldi committed:
(refactor) Create helper sum function

Function simply adds an array of numbers together.

cgrinaldi committed:
(refactor) Use helper sum function in calcAverageRating()

cgrinaldi committed:
(feat) Create createSamplingDists()

Function takes in an array of weighted problems (tuples) and
returns an array of tuples where tuple[0] contains the value
that we will use to sample from.

cgrinaldi committed:
(feat) Create getProblem() and wire up service correctly

This is the method that is exposed by our LearningAlgo service.
It simply selects a problem from the array of problems that a
user has, using weights calculated by previous performance, time
since last seen, and if the user missed the most recent attempt or
not.

cgrinaldi committed:
(refactor) Change dummy rating data

Instead of having the rating data be a string ("easy", "hard", etc.),
it makes more sense to have it on a numerical scale between 0 and 1.

cgrinaldi committed:
(refactor) Change editorController to use getProblem()

Prior to this, editorController was using the LearningAlgo service
to get an array of problems. Now, the controller will simply
call getProblem(), and the learning algorithm will handle the
delivering of a single problem object.

cgrinaldi committed:
(doc) Add TODO about optimizing search

mosse commented:
Looks fantastic! Nicely done. No worries at all about changing the references. Happy to merge given Garrett is traveling today. (Hope that's cool Garrett...)

cgrinaldi committed:
(feat) Remove problem from consideration after call to getProblem()

Added a helper function that will remove a problem after the user sees
it. This way, in a given session, the user does not see the same
problem over and over again.

Currently, the function assumes that each problem can be
uniquely identified by its title attribute.

cgrinaldi committed:
(fix) Return null if no new problems present for the user
